src/coord/rpnet.c

```diff
@@ -73,27 +74,231 @@ static RSValue *MRReply_ToValue(MRReply *r) {
-    // TODO: could be replaced with a query specific configuration
-    if (!MR_ManuallyTriggerNextIfNeeded(nc->it, clusterConfig.cursorReplyThreshold)) {
-      // No more replies
-      RPNet_resetCurrent(nc);
-      return 0;
+  // Wait for all shards' first responses before returning any results
+  // This ensures accurate total_results from the start
+  MRReply *root = NULL;
+  bool timedOut = false;
+  bool errorFound = false;
+  if (nc->shardResponseBarrier && !nc->waitedForAllShards) {
+    const size_t numShards = nc->shardResponseBarrier->numShards;
+
+    // Collect replies until all shards have sent their first response
+    // Also respect the query timeout to avoid blocking indefinitely
+    while (atomic_load(&nc->shardResponseBarrier->numResponded) < numShards) {
+      // Check for timeout to avoid blocking indefinitely
+      if (nc->areq && nc->areq->sctx && TimedOut(&nc->areq->sctx->time.timeout)) {
+        timedOut = true;
+        break;
+      }
+      // Calculate absolute timeout for MRIterator_NextWithTimeout
+      struct timespec absTimeout;
+      struct timespec *timeoutPtr = calculateAbsTimeout(nc, &absTimeout);
+
+      if (nc->cmd.forCursor) {
+        if (!MR_ManuallyTriggerNextIfNeeded(nc->it, clusterConfig.cursorReplyThreshold)) {
+          break;  // No more replies available
+        }
+      }
+
+      // Get next reply with timeout
+      bool nextTimedOut = false;
+      MRReply *reply = MRIterator_NextWithTimeout(nc->it, timeoutPtr, &nextTimedOut);
+      if (reply == NULL) {
+        if (nextTimedOut) {
+          timedOut = true;
+        }
+        break;  // No more replies or timed out
+      }
+
+      // Store reply for later processing
+      if (!nc->pendingReplies) {
+        nc->pendingReplies = array_new(MRReply *, numShards);
+      }
+      array_append(nc->pendingReplies, reply);
+
+      // Check for errors
+      if (shardResponseBarrier_HandleError(nc)) {
+        root = nc->current.root;
+        errorFound = true;
+        break;
+      }
+    }
+
+    // Mark that we've waited (even if not all shards responded due to time out - to avoid infinite loop)
+    nc->waitedForAllShards = true;
+
+    // Handle timeout or not enough shards responded
+    if (timedOut) {
+      shardResponseBarrier_HandleTimeout(nc);
+      root = NULL;
+    }
+    shardResponseBarrier_UpdateTotalResults(nc);
+  }
+
+  if (!errorFound) {
+    // First, return any pending replies collected during the wait
+    if (nc->pendingReplies && array_len(nc->pendingReplies) > 0) {
+      // Pop the first pending reply
+      root = nc->pendingReplies[0];
+      array_del(nc->pendingReplies, 0);
+    } else {
+      // No pending replies, get from channel
+      if (nc->cmd.forCursor) {
+        if (!MR_ManuallyTriggerNextIfNeeded(nc->it, clusterConfig.cursorReplyThreshold)) {
+          RPNet_resetCurrent(nc);
+          return 0;
+        }
+      }
+      root = MRIterator_Next(nc->it);
     }
   }
```

Comment:
Can we avoid polluting code unrelated to the feature? What's the expected behavior when `errorFound` is true? Can we handle that in place and not here/later?

Reply:
Fixed [c2ea61d](https://github.com/RediSearch/RediSearch/pull/7202/commits/c2ea61dd8dd847705688b23b5705123a15b60432)